File: src/linear_solver.f90
module linear_solver
    use kinds, only: dp
    use errors, only: require
    use matrix_operations, only: lu_factor, forward_substitution, backward_substitution, vector_norm2, vector_dot
    implicit none
contains
    subroutine solve_linear_system(A, b, x, info)
        real(dp), dimension(:,:), intent(in) :: A
        real(dp), dimension(:), intent(in) :: b
        real(dp), dimension(:), intent(out) :: x
        integer, intent(out) :: info
        real(dp), allocatable :: L(:,:), U(:,:), y(:)
        integer, allocatable :: piv(:)
        integer :: n
        n = size(A,1)
        call require(size(A,2) == n, 'solve_linear_system: A not square')
        call require(size(b) == n, 'solve_linear_system: b size mismatch')
        call require(size(x) == n, 'solve_linear_system: x size mismatch')

        allocate(L(n,n), U(n,n), y(n), piv(n))
        call lu_factor(A, L, U, piv, info)
        if (info /= 0) then
            x = 0.0_dp
            deallocate(L, U, y, piv)
            return
        end if
        call apply_pivot(b, piv, y)
        call forward_substitution(L, y, y)
        call backward_substitution(U, y, x)
        deallocate(L, U, y, piv)
    end subroutine solve_linear_system

    subroutine apply_pivot(b, piv, bp)
        real(dp), dimension(:), intent(in) :: b
        integer, dimension(:), intent(in) :: piv
        real(dp), dimension(:), intent(out) :: bp
        integer :: n, i
        n = size(b)
        call require(size(piv) == n, 'apply_pivot: size mismatch')
        call require(size(bp) == n, 'apply_pivot: output size mismatch')
        do i = 1, n
            bp(i) = b(piv(i))
        end do
    end subroutine apply_pivot

    subroutine jacobi_solver(A, b, x, max_iter, tol, iterations, converged)
        real(dp), dimension(:,:), intent(in) :: A
        real(dp), dimension(:), intent(in) :: b
        real(dp), dimension(:), intent(inout) :: x
        integer, intent(in) :: max_iter
        real(dp), intent(in) :: tol
        integer, intent(out) :: iterations
        logical, intent(out) :: converged
        integer :: n, i, j
        real(dp) :: sigma, diff
        real(dp), allocatable :: x_new(:)

        n = size(A,1)
        call require(size(A,2) == n, 'jacobi_solver: A not square')
        call require(size(b) == n .and. size(x) == n, 'jacobi_solver: size mismatch')

        allocate(x_new(n))
        converged = .false.
        do iterations = 1, max_iter
            do i = 1, n
                sigma = 0.0_dp
                do j = 1, n
                    if (j /= i) sigma = sigma + A(i,j) * x(j)
                end do
                call require(A(i,i) /= 0.0_dp, 'jacobi_solver: zero diagonal element')
                x_new(i) = (b(i) - sigma) / A(i,i)
            end do
            diff = vector_norm2(x_new - x)
            x = x_new
            if (diff < tol) then
                converged = .true.
                exit
            end if
        end do
        deallocate(x_new)
    end subroutine jacobi_solver

    subroutine gauss_seidel_solver(A, b, x, max_iter, tol, iterations, converged)
        real(dp), dimension(:,:), intent(in) :: A
        real(dp), dimension(:), intent(in) :: b
        real(dp), dimension(:), intent(inout) :: x
        integer, intent(in) :: max_iter
        real(dp), intent(in) :: tol
        integer, intent(out) :: iterations
        logical, intent(out) :: converged
        integer :: n, i, j
        real(dp) :: sigma, diff
        real(dp), allocatable :: x_old(:)

        n = size(A,1)
        call require(size(A,2) == n, 'gauss_seidel_solver: A not square')
        call require(size(b) == n .and. size(x) == n, 'gauss_seidel_solver: size mismatch')

        allocate(x_old(n))
        converged = .false.
        do iterations = 1, max_iter
            x_old = x
            do i = 1, n
                sigma = 0.0_dp
                do j = 1, n
                    if (j /= i) sigma = sigma + A(i,j) * x(j)
                end do
                call require(A(i,i) /= 0.0_dp, 'gauss_seidel_solver: zero diagonal element')
                x(i) = (b(i) - sigma) / A(i,i)
            end do
            diff = vector_norm2(x - x_old)
            if (diff < tol) then
                converged = .true.
                exit
            end if
        end do
        deallocate(x_old)
    end subroutine gauss_seidel_solver

    subroutine conjugate_gradient_solver(A, b, x, max_iter, tol, iterations, converged)
        real(dp), dimension(:,:), intent(in) :: A
        real(dp), dimension(:), intent(in) :: b
        real(dp), dimension(:), intent(inout) :: x
        integer, intent(in) :: max_iter
        real(dp), intent(in) :: tol
        integer, intent(out) :: iterations
        logical, intent(out) :: converged
        real(dp), allocatable :: r(:), p(:), Ap(:)
        real(dp) :: alpha, beta, rs_old, rs_new
        integer :: n

        n = size(A,1)
        call require(size(A,2) == n, 'conjugate_gradient_solver: A not square')
        call require(size(b) == n .and. size(x) == n, 'conjugate_gradient_solver: size mismatch')

        allocate(r(n), p(n), Ap(n))
        r = b - matmul(A, x)
        p = r
        rs_old = vector_dot(r, r)
        converged = .false.
        do iterations = 1, max_iter
            Ap = matmul(A, p)
            alpha = rs_old / vector_dot(p, Ap)
            x = x + alpha * p
            r = r - alpha * Ap
            rs_new = vector_dot(r, r)
            if (sqrt(rs_new) < tol) then
                converged = .true.
                exit
            end if
            beta = rs_new / rs_old
            p = r + beta * p
            rs_old = rs_new
        end do
        deallocate(r, p, Ap)
    end subroutine conjugate_gradient_solver
end module linear_solver

File: python_solver/linear_solver.py
from __future__ import annotations

from typing import List, Tuple

from python_solver.utils import dot, norm2


def lu_factor(a: List[List[float]]) -> Tuple[List[List[float]], List[List[float]], List[int], int]:
    n = len(a)
    l = [[0.0 for _ in range(n)] for _ in range(n)]
    u = [row[:] for row in a]
    piv = list(range(n))
    info = 0

    for k in range(n):
        pivot_row = max(range(k, n), key=lambda i: abs(u[i][k]))
        if u[pivot_row][k] == 0.0:
            info = k + 1
            break
        if pivot_row != k:
            u[k], u[pivot_row] = u[pivot_row], u[k]
            l[k], l[pivot_row] = l[pivot_row], l[k]
            piv[k], piv[pivot_row] = piv[pivot_row], piv[k]
        l[k][k] = 1.0
        for i in range(k + 1, n):
            l[i][k] = u[i][k] / u[k][k]
            for j in range(k, n):
                u[i][j] -= l[i][k] * u[k][j]
    return l, u, piv, info


def forward_substitution(l: List[List[float]], b: List[float]) -> List[float]:
    n = len(l)
    y = [0.0 for _ in range(n)]
    for i in range(n):
        y[i] = b[i] - sum(l[i][j] * y[j] for j in range(i))
        y[i] /= l[i][i]
    return y


def backward_substitution(u: List[List[float]], y: List[float]) -> List[float]:
    n = len(u)
    x = [0.0 for _ in range(n)]
    for i in range(n - 1, -1, -1):
        x[i] = y[i] - sum(u[i][j] * x[j] for j in range(i + 1, n))
        x[i] /= u[i][i]
    return x


def solve_linear_system(a: List[List[float]], b: List[float]) -> Tuple[List[float], int]:
    l, u, piv, info = lu_factor(a)
    if info != 0:
        return [0.0 for _ in b], info
    bp = [b[piv[i]] for i in range(len(b))]
    y = forward_substitution(l, bp)
    x = backward_substitution(u, y)
    return x, 0


def jacobi(a: List[List[float]], b: List[float], max_iter: int, tol: float) -> Tuple[List[float], int, bool]:
    n = len(a)
    x = [0.0 for _ in range(n)]
    for iteration in range(1, max_iter + 1):
        x_new = [0.0 for _ in range(n)]
        for i in range(n):
            sigma = sum(a[i][j] * x[j] for j in range(n) if j != i)
            x_new[i] = (b[i] - sigma) / a[i][i]
        diff = norm2([x_new[i] - x[i] for i in range(n)])
        x = x_new
        if diff < tol:
            return x, iteration, True
    return x, max_iter, False


def gauss_seidel(a: List[List[float]], b: List[float], max_iter: int, tol: float) -> Tuple[List[float], int, bool]:
    n = len(a)
    x = [0.0 for _ in range(n)]
    for iteration in range(1, max_iter + 1):
        x_old = x[:]
        for i in range(n):
            sigma = sum(a[i][j] * x[j] for j in range(n) if j != i)
            x[i] = (b[i] - sigma) / a[i][i]
        diff = norm2([x[i] - x_old[i] for i in range(n)])
        if diff < tol:
            return x, iteration, True
    return x, max_iter, False


def conjugate_gradient(a: List[List[float]], b: List[float], max_iter: int, tol: float) -> Tuple[List[float], int, bool]:
    n = len(a)
    x = [0.0 for _ in range(n)]
    r = [b[i] - dot(a[i], x) for i in range(n)]
    p = r[:]
    rs_old = dot(r, r)
    for iteration in range(1, max_iter + 1):
        ap = [dot(a[i], p) for i in range(n)]
        alpha = rs_old / dot(p, ap)
        x = [x[i] + alpha * p[i] for i in range(n)]
        r = [r[i] - alpha * ap[i] for i in range(n)]
        rs_new = dot(r, r)
        if rs_new ** 0.5 < tol:
            return x, iteration, True
        beta = rs_new / rs_old
        p = [r[i] + beta * p[i] for i in range(n)]
        rs_old = rs_new
    return x, max_iter, False
